### PR TITLE
[sharktank] General implementation for trivially replicable ops

### PR DESCRIPTION
**Pull request by sogartar**
The majority of operations have essentially the same replication mechanism across device. We call the underlying unsharded op with the corresponding shards and then construct a replicated tensor form the resulting shards.

This change solves this in general way to not have to write the same replication pattern for all ops.

### CODE REVIEW

**Comment by sogartar** [comment]
This causes a failure when tracing with torch 2.3, but it seems we are going to drop 2.3. I will open a PR for that shortly, but other than that the PR is good for review.

File: sharktank/sharktank/ops/utils.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        fn=reduce_to_list_of_tensors_fn,
+        trees=per_shard_results,
+        is_leaf=is_leaf,
+        initial=reduce_initial,
+    )
+
+    def make_replicated_tensor_if_tensor_collection(x: Any) -> Any:
+        if _is_tensor_collection(x):
+            return ReplicatedTensor(ts=x, devices=first_replicated_tensor_arg.devices)
+        return x
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/utils.py:94]: This is only doing some of the work in `sharded_wrap_override`. And now some ops are going through `sharded_wrap_override` and others are not (even for ReplicatedTensors). You shouldn't need to specify `devices` here, and instead be wrapping so that we call `sharded_wrap_override`.

File: sharktank/sharktank/ops/utils.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    def reduce_to_list_of_tensors_fn(a: Any, b: Any) -> Any:
+        if not is_any_tensor(b):
+            return a
+        a.append(b)
+        return a
+
+    def tensor_to_empty_list(x: Any) -> list:
+        if is_any_tensor(x):
+            return []
+        return x
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/utils.py:77]:
```suggestion
        return [] if is_any_tensor(x) else x
```

File: sharktank/sharktank/ops/utils.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        for per_shard_args, per_shard_kwargs in per_shard_args_and_kwargs
+    ]
+
+    def reduce_to_list_of_tensors_fn(a: Any, b: Any) -> Any:
+        if not is_any_tensor(b):
+            return a
+        a.append(b)
+        return a
+
+    def tensor_to_empty_list(x: Any) -> list:
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/utils.py:74]: The type hint for the return isn't correct. It should be either `list | Any` or just `Any`.

File: sharktank/sharktank/ops/utils.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    if isinstance(v, ReplicatedTensor):
+        return v.shards[shard_index]
+    return v
+
+
+def _is_leaf(v: Any) -> bool:
+    from sharktank.types import (
+        is_any_tensor,
+    )
```
Comment (Alex-Vasile) [inline sharktank/sharktank/ops/utils.py:161]: Is the comma after `is_any_tensor` needed? We don't do this with other single imports, and it's causing black to format this over 3 lines instead of 1, making both this and the outer function much longer than necessary. Same comment for many of the other one-line functions

File: sharktank/sharktank/utils/tree.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 from typing import Dict, Any, Callable, Mapping, Iterable, Sequence, Union
 from collections.abc import Mapping as MappingABC, Iterable as IterableABC
+import functools
```
Comment (Alex-Vasile) [inline sharktank/sharktank/utils/tree.py:9]: You'll want someone else to review this file, since I haven't worked with the tree functionality before.

File: sharktank/sharktank/ops/utils.py (hunk 1 above)
Comment (Alex-Vasile) [reply sharktank/sharktank/ops/utils.py:94]: That this isn't failing anything means that we need more tests, including a few negative ones, inside `pipeline_parallized_test.py`.

For example, adding the following test to `pipeline_parallized_test.py::CatTest`, will throw an error since the devices of the args aren't being checked:
```python
    def testCatReplicatedShouldFail(self):
        """Concatenation of replicated tensors."""
        shard_count = 2
        cat_dim = 0
        # Done to ensure no overlap with default devices so incorrect placement is caught
        devices = tuple(range(shard_count, 2 * shard_count))

        a = torch.rand(5, 4, dtype=torch.float32)
        b = torch.rand(3, 4, dtype=torch.float32)

        sharded_a = ReplicatedTensor(ts=a, shard_count=shard_count, devices=devices)
        sharded_b = ReplicatedTensor(ts=b, shard_count=shard_count, devices=tuple(range(shard_count)))

        try:
            ops.cat([sharded_a, sharded_b], dim=cat_dim)
        except ValueError:
            return  # Should throw an error since the tensors are not on the same devices
        raise AssertionError("Should have thrown an error since the tensors are not on the same devices")
```

**Review by Alex-Vasile — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/utils/tree.py (hunk 5 above)
Comment (sogartar) [reply sharktank/sharktank/utils/tree.py:9]: I don't think anyone else is. This is some general functionality for tree manipulation. If it is not clear I can add more documentation.

File: sharktank/sharktank/ops/utils.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    When collecting the results all non-tensor results are taken from the first device.
+    """
+
+    from sharktank.types import (
+        ReplicatedTensor,
+        is_any_tensor,
+    )
+
+    def is_leaf(v: Any) -> bool:
+        return is_any_tensor(v) or tree.is_leaf_default(v)
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/ops/utils.py]: Removed it.

File: sharktank/sharktank/types/tensors.py [hunk 7]
```diff
@@ -1459,6 +1460,10 @@ def clone(self, **kwargs) -> "UnreducedTensor":
         return UnreducedTensor(**kwargs)
 
 
+def is_any_tensor(x: Any) -> bool:
+    return isinstance(x, InferenceTensor) or isinstance(x, torch.Tensor)
```
[the remark this answers is not in the document]
Comment (sogartar) [reply sharktank/sharktank/types/tensors.py]: Done.

File: sharktank/sharktank/ops/utils.py (hunk 1 above)
Comment (sogartar) [reply sharktank/sharktank/ops/utils.py]: I refactored a bit such that the trivially replicated overrides happen inside `sharded_impls.py`. I may come back at some point and refactor this composition. We can't declare right now ops outside of `signatures.py`. For example we can't have tests declare ops.

File: sharktank/sharktank/ops/utils.py (hunk 3 above)
Comment (sogartar) [reply sharktank/sharktank/ops/utils.py]: Done.

File: sharktank/tests/ops/sharded_test.py [hunk 8]
```diff
@@ -295,6 +299,26 @@ def testCatNonSplitDim(self):
         actual_result = ops.cat([sharded_a, sharded_b], dim=cat_dim)
         assert expected_result.is_deep_equal(actual_result, compare_name=False)
 
+    def testCatReplicatedShouldFail(self):
```
Comment (Alex-Vasile) [inline sharktank/tests/ops/sharded_test.py:302]: This probably belongs in `pipeline_parallelized_test.py` since that's where all the pp-related ops tests currently are.

File: sharktank/sharktank/utils/tree.py [hunk 9]
```diff
@@ -14,59 +15,157 @@
...
+    dict_type: type = dict,
+    sequence_type: type = tuple,
+) -> Tree:
+    """Reduce a list of trees such that elements with the same path in the tree are
+    aligned.
+
+    E.g.
+    ```
+    {"a": a1, {"b": b1}}
+    {"a": a2, {"b": b2}}
```
Comment (Alex-Vasile) [inline sharktank/sharktank/utils/tree.py:122]: I don't understand what this example is showing. Is it complete, or what am I missing?

**Review by Alex-Vasile — APPROVED** [review]
Two minor comments, but good to go.

File: sharktank/tests/ops/sharded_test.py (hunk 8 above)
Comment (sogartar) [reply sharktank/tests/ops/sharded_test.py]: I moved it.
